op-service/signer/xlayer_auth.go

```diff
@@ -0,0 +1,209 @@
+package signer
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/md5"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strings"
+)
+
+const (
+	algoSha256      = "sha256"
+	algoMd5         = "md5"
+	headerSignKey   = "sign"
+	headerAccessKey = "accessKey"
+)
+
+// auth is SecretKey or AccessKey is empty then no auth is performed
+func (c *XLayerRemoteClient) addAuth(req *http.Request) error {
+	if c.config.SecretKey == "" || c.config.AccessKey == "" {
+		return nil
+	}
+	req.Header[headerAccessKey] = []string{c.config.AccessKey}
+	// Generate the signature
+	signature, err := c.genAuth(req, algoSha256)
+	if err != nil {
+		return fmt.Errorf("failed to generate signature: %v", err)
+	}
+	req.Header[headerSignKey] = []string{signature}
+
+	return nil
+}
+
+func (c *XLayerRemoteClient) genAuth(req *http.Request, algorithm string) (string, error) {
+	params := req.URL.Query()
+
+	treeMap := make(map[string][]string)
+	for _, v := range params {
+		var key string
+		for _, vv := range v {
+			key += vv
+		}
+		treeMap[key] = v
+	}
+
+	var body strings.Builder
+	if req.Body != nil {
+		readCloser, err := req.GetBody()
+		if err != nil {
+			return "", fmt.Errorf("get body error: %v", err)
+		}
+		defer readCloser.Close()
+
+		buffer := make([]byte, 0)
+		// Read the request body into the 'buffer' variable
+		_, err = io.Copy(&bufferWriter{&buffer}, readCloser)
+		if err != nil {
+			return "", fmt.Errorf("read body error: %v", err)
+		}
+
+		// Append the body if present
+		if len(buffer) != 0 {
+			body.Write(buffer)
+		}
+	}
+
+	// Calculate the hash based on the selected algorithm
+	return c.generateSignature(treeMap, body.String(), algorithm)
+}
+
+func (c *XLayerRemoteClient) generateSignature(treeMap map[string][]string, body, algorithm string) (string, error) {
+	// Sort the map by values
+	var keys []string
+	for key := range treeMap {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+
+	// Construct the content string
+	var content strings.Builder
+	for _, key := range keys {
+		value := treeMap[key]
+		for _, v := range value {
+			content.WriteString(v)
+		}
+	}
+	// Append the body if present
+	if body != "" {
+		content.WriteString(body)
+	}
+
+	// Calculate the hash based on the selected algorithm
+	var hash []byte
+	switch strings.ToLower(algorithm) {
+	case "sha256":
+		hash = signBySha256(content.String())
+	case "md5":
+		hash = signByMd5(content.String())
```

Comment:
MD5 has been compromised and should no longer be used in projects. Will it be replaced with a more secure algorithm like SHA-256 in the future?